src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
+    this.children = config.children.map(
+        config => new ModulePluginWrapper(logger, config, traceConfig));
+    // Eagerly load core plugins into memory.
+    // This prevents issues related to circular dependencies.
+    this.children.forEach(child => child.getPluginExportedValue());
+  }
+
+  /**
+   * Returns whether the given version of the module is supported by this
+   * plugin. This may load the plugin into memory.
+   * @param version A semver version string.
+   */
+  isSupported(version: string): boolean {
+    return this.children.some(child => child.isSupported(version));
+  }
+  /**
+   * Call unpatch methods when they were provided.
+   */
+  unapplyAll(): void {
+    this.children.forEach(child => child.unapplyAll());
+  }
+  /**
+   * Applies this object's underlying plugin patches to a file, returning the
+   * patched or intercepted value.
+   * @param moduleExports The module exports of the file.
+   * @param file The file path, relative to the module root.
+   * @param version The module version.
+   */
+  applyPlugin<T>(moduleExports: T, file: string, version: string): T {
+    return this.children.reduce((exportedValue, child) => {
+      return child.applyPlugin(exportedValue, file, version);
+    }, moduleExports);
   }
-  activated = true;
+}
+
+// States for the Plugin Loader
+export enum PluginLoaderState {
+  NO_HOOK,
+  ACTIVATED,
+  DEACTIVATED
+}
+
+/**
+ * A class providing functionality to hook into module loading and apply
+ * plugins to enable tracing.
+ */
+export class PluginLoader {
+  // Key on which core modules are stored.
+  static readonly CORE_MODULE = '[core]';
+  // The function to call to register a require hook.
+  private enableRequireHook: (onRequire: hook.OnRequireFn) => void;
+  // A logger.
+  private logger: Logger;
+  // A map mapping module names to their respective plugins.
+  private pluginMap: Map<string, PluginWrapper> = new Map();
+  // A map caching version strings for a module based on their base path.
+  private moduleVersionCache: Map<string, string|null> = new Map();
+  // The current state of the plugin loader.
+  private internalState: PluginLoaderState = PluginLoaderState.NO_HOOK;
+
+  /**
+   * Constructs a new PluginLoader instance.
+   * @param logger The logger to use.
+   * @param config The configuration for this instance.
+   */
+  constructor(logger: Logger, config: PluginLoaderConfig) {
+    this.logger = logger;
 
-  logger_ = logger;
+    const nonCoreModules: string[] = [];
+    // Initialize ALL of the PluginWrapper objects here.
+    // See CorePluginWrapper docs for why core modules are processed
+    // differently.
+    const coreWrapperConfig: CorePluginWrapperOptions = {children: []};
+    Object.keys(config.plugins).forEach(key => {
+      const value = config.plugins[key];
+      // Core module plugins share a common key.
+      const coreModule = builtinModules.indexOf(key) !== -1 ||
+          key === PluginLoader.CORE_MODULE;
```

Comment:
Isn't it better to do the latter comparison first because `indexOf()` is O(N)?

Reply:
Yep -- done